packages/cardano-services/src/util/TypeormProvider/TypeormProvider.ts

```diff
@@ -1,26 +1,34 @@
 import { HealthCheckResponse, Milliseconds, Provider } from '@cardano-sdk/core';
+import { InMemoryCache } from '../../InMemoryCache';
 import { TypeormService, TypeormServiceDependencies } from '../TypeormService';
 import { skip } from 'rxjs';
 
-export type TypeormProviderDependencies = Omit<TypeormServiceDependencies, 'connectionTimeout'>;
+export type TypeormProviderDependencies = Omit<TypeormServiceDependencies, 'connectionTimeout'> & {
+  healthCheckCache: InMemoryCache;
+};
 
 const unhealthy = { ok: false, reason: 'Provider error' };
 
 export abstract class TypeormProvider extends TypeormService implements Provider {
   health: HealthCheckResponse = { ok: false, reason: 'not started' };
+  healthCheckCache: InMemoryCache;
 
   constructor(name: string, dependencies: TypeormProviderDependencies) {
     super(name, { ...dependencies, connectionTimeout: Milliseconds(1000) });
     // We skip 1 to omit the initial null value of the subject
     this.dataSource$.pipe(skip(1)).subscribe((dataSource) => {
       this.health = dataSource ? { ok: true } : unhealthy;
     });
+
+    this.healthCheckCache = dependencies.healthCheckCache;
   }
 
   async healthCheck(): Promise<HealthCheckResponse> {
     if (this.state === 'running')
       try {
-        await this.withDataSource((dataSource) => dataSource.query('SELECT 1'));
+        await this.healthCheckCache.get('typeorm_db', () =>
+          this.withDataSource((dataSource) => dataSource.query('SELECT 1'))
```

Comment:
Sorry, I can't understand how possible timeouts here are caught.

Reply:
@iccicci The timeout will be thrown inside `.withDataSource` and caught by the `try-block` since we do `await` the returned `promise` inside it. Please let me know if I missed the point.

I extended the existing test case with a thrown error in f66478c45e886ceb793bce21c2be290647837d50

Reply:
I'm wondering what happens if the function `() => this.withDataSource((dataSource) => dataSource.query('SELECT 1'))` (or the `() => this.#checkHealth()` in `NodeTxSubmitProvider.ts`) takes (for example) 5 minutes.

Will the health check take 5 minutes to become unhealthy as well?
Due to the nature of `InMemoryCache`, in case of expired cache, the `get` method waits till the resolution (or rejection) of the promise; could it be that in case of expired cache AND a long time required for the callback make the health check to take 5 minutes to respond?

Reply:
Good point @iccicci .
I changed the implementation, so the cache will be updated once the promise is completed. Except the first time calling, the caller will get already completed promises.
Here: a64610fe47ebff27ca351e56cc0dcaa1f55b4569